Yosemite/Yosemite/Stores/TaxClassesStore.swift

```diff
@@ -0,0 +1,153 @@
+import Foundation
+import Networking
+import Storage
+
+
+// MARK: - TaxClassesStore
+//
+public class TaxClassesStore: Store {
+
+    private lazy var sharedDerivedStorage: StorageType = {
+        return storageManager.newDerivedStorage()
+    }()
+
+    /// Registers for supported Actions.
+    ///
+    override public func registerSupportedActions(in dispatcher: Dispatcher) {
+        dispatcher.register(processor: self, for: ProductAction.self)
+    }
+
+    /// Receives and executes Actions.
+    ///
+    override public func onAction(_ action: Action) {
+        guard let action = action as? TaxClassesAction else {
+            assertionFailure("TaxClassesStore received an unsupported action")
+            return
+        }
+
+        switch action {
+        case .retriveTaxClasses(let siteID, let onCompletion):
+            retriveTaxClasses(siteID: siteID, onCompletion: onCompletion)
+        case .resetStoredTaxClasses(let onCompletion):
+            resetStoredTaxClasses(onCompletion: onCompletion)
+        case .requestMissingTaxClasses(let product, let onCompletion):
+            requestMissingTaxClasses(for: product, onCompletion: onCompletion)
+        }
+
+    }
+}
+
+
+// MARK: - Services!
+//
+private extension TaxClassesStore {
+
+    /// Retrieve and synchronizes the Tax Classes associated with a given Site ID (if any!).
+    ///
+    func retriveTaxClasses(siteID: Int, onCompletion: @escaping ([TaxClass]?, Error?) -> Void) {
+        let remote = TaxClassesRemote(network: network)
+
+        remote.loadAllTaxClasses(for: Int64(siteID)) { [weak self] (taxClasses, error) in
+            guard let taxClasses = taxClasses else {
+                onCompletion(nil, error)
+                return
+            }
+
+            self?.upsertStoredTaxClassesInBackground(readOnlyTaxClasses: taxClasses) {
+                onCompletion(taxClasses, nil)
+            }
+        }
+    }
+
+    /// Deletes all of the Stored TaxClasses.
+    ///
+    func resetStoredTaxClasses(onCompletion: () -> Void) {
+        let storage = storageManager.viewStorage
+        storage.deleteAllObjects(ofType: Storage.TaxClass.self)
+        storage.saveIfNeeded()
+        DDLogDebug("Tax Classes deleted")
+
+        onCompletion()
+    }
+
+    /// Synchronizes the Tax Class found in a specified Product.
+    ///
+    func requestMissingTaxClasses(for product: Product, onCompletion: @escaping ([TaxClass]?, Error?) -> Void) {
```

Comment:
since a `Product` has up to one tax class, do you think we could return just one optional `TaxClass` in the completion block instead of an array of tax classes?

Reply:
I missed it, but it makes sense! Thanks for noticing it 🙇